lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs

```diff
@@ -0,0 +1,810 @@
+--
+-- | Address type       | Payment Credential | Stake Credential | Header, first nibble |
+-- |--------------------+--------------------+------------------+----------------------|
+-- | Base address       | keyhash            | keyhash          |                 0000 |
+-- |                    | scripthash         | keyhash          |                 0001 |
+-- |                    | keyhash            | scripthash       |                 0010 |
+-- |                    | scripthash         | scripthash       |                 0011 |
+-- | Pointer address    | keyhash            | ptr              |                 0100 |
+-- |                    | scripthash         | ptr              |                 0101 |
+-- | Enterprise address | keyhash            | -                |                 0110 |
+-- |                    | scripthash         | 0                |                 0111 |
+-- | Bootstrap address  | keyhash            | -                |                 1000 |
+-- | Stake address      | -                  | keyhash          |                 1110 |
+-- |                    | -                  | scripthash       |                 1111 |
+-- | Future formats     | ?                  | ?                |            1001-1101 |
+--
+-- We represent these types of addresses with the following data types:
+
+-- | The type of the address.
+data AddressType
+    = BaseAddress Credential Credential
+    | PointerAddress Credential
+    | EnterpriseAddress Credential
+    | StakeAddress Credential
+    | BootstrapAddress
+    -- ^ A Bootstrap (a.k.a. Byron) address
+    deriving (Eq, Show)
+
+-- | The type of the credential used in an address.
+data Credential
+    = CredentialKeyHash
+    | CredentialScriptHash
+    deriving (Eq, Show)
+
+-- To parse the address type, we can inspect the first four bits (nibble) of the
+-- address:
+
+-- | Get an AddressType from a binary stream.
+getHeader :: B.Get AddressType
+getHeader = do
+    headerAndNetwork <- B.getWord8
+    let
+        header =
+            -- Mask for just the address type nibble
+            headerAndNetwork .&. 0b11110000
+    case header of
+        0b00000000 -> pure $ BaseAddress CredentialKeyHash CredentialKeyHash
+        0b00010000 -> pure $ BaseAddress CredentialScriptHash CredentialKeyHash
+        0b00100000 -> pure $ BaseAddress CredentialKeyHash CredentialScriptHash
+        0b00110000 -> pure $ BaseAddress CredentialScriptHash CredentialScriptHash
+        0b01000000 -> pure $ PointerAddress CredentialKeyHash
+        0b01010000 -> pure $ PointerAddress CredentialScriptHash
+        0b01100000 -> pure $ EnterpriseAddress CredentialKeyHash
+        0b01110000 -> pure $ EnterpriseAddress CredentialScriptHash
+        0b10000000 -> pure BootstrapAddress
+        0b11100000 -> pure $ StakeAddress CredentialKeyHash
+        0b11110000 -> pure $ StakeAddress CredentialScriptHash
+        _          -> fail "Unknown address type."
```

Comment:
Perhaps we can separate out the following responsibilities:
    
- defining the bidirectional mapping between an address type and its corresponding header nibble
  (this can be expressed in terms of **`Word4`**)
- reading or modifying the address type of an address

I've made a commit to demonstrate the idea here: 973ecc67a2ffdefcf9008f189b1d1747d752e9d9

What do you think?

Reply:
That's great!